### PR TITLE
Handle generated columns on ghostferry write operations.

### PR DESCRIPTION
**Pull request by proton-lisandro-pin**
This PR modifies all `INSERT` logic so virtual (a.k.a generated) MySQL columns are not written into, which would otherwise break the ferrying process.

See also https://github.com/Shopify/ghostferry/issues/338.

### CODE REVIEW

**Comment by proton-lisandro-pin** [comment]
_(second)_ Friendly ping? :smile:

**Review by grodowski — CHANGES_REQUESTED** [review]
Thanks for the contribution @proton-lisandro-pin! Reviewed together with @driv3r. 

Instead of filtering in `RowBatch`, we could filter generated columns in `table_schema_cache.LoadTables`, which would look similarly to a [recent index visibility change](https://github.com/Shopify/ghostferry/pull/395/changes#diff-2ff6bf0ca67c3788907ff97b3ecc86c3ad30011baed6d18df8b50b9f7a981352R192-R199). This seems cleaner and more performant, since the row batch approach would recompute at every iteration.

For completeness, the filtering should reject both `IsVisible` and `IsStored` since these are [treated separately](https://github.com/go-mysql-org/go-mysql/blob/22706568523cd1a8aa8b52fbdd03c3a570f36f5c/schema/schema.go#L156-L159) by `go-mysql`.

The other issue is that `dml_events.go` should also consider syntax changes for updates and deletes (I think we observed integration test failures due to this).

We'd suggest to update `test/helpers/db_helper.rb` to include generated columns to the setup create tables as a starting point.

**Comment by proton-lisandro-pin** [comment]
@driv3r @grodowski Thanks for the feedback!

It's been a while since i first uploaded this PR :man_facepalming: i'll rework as suggested and ping back once it's ready.

**Review by driv3r — CHANGES_REQUESTED** [review]
sorry for delay, sick days, we should have tighter loop from now on

**Review by driv3r — CHANGES_REQUESTED** [review]
I think we are almost there :)

File: table_schema_cache.go [hunk 1]
```diff
@@ -46,6 +46,89 @@ type TableSchema struct {
...
+	for _, col := range t.Columns {
+		if IsColumnGenerated(&col) {
+			generated += 1
+		}
+	}
+
+	return len(t.Columns), generated, len(t.Columns) - generated
+}
+
+// Returns a list of all non-generated column nams for a TableSchema, in schema order.
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply table_schema_cache.go]: Done.

File: table_schema_cache.go [hunk 2]
```diff
@@ -46,6 +46,89 @@ type TableSchema struct {
 	rowMd5Query string
 }
 
+// IsColumnGenerated evaluates wheter a go_myslq.schema.TableColumn is generated or not.
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply table_schema_cache.go]: Done.

File: table_schema_cache.go [hunk 3]
```diff
@@ -46,6 +46,89 @@ type TableSchema struct {
...
+func IsColumnGenerated(tc *schema.TableColumn) bool {
+	return tc.IsVirtual || tc.IsStored
+}
+
+// IsColumnIndexGenerated evaluetes whether a TableSchema column is generated, by index.
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply table_schema_cache.go]: Done.

File: table_schema_cache.go [hunk 4]
```diff
@@ -151,6 +235,19 @@ func MaxPaginationKeys(db *sql.DB, tables []*TableSchema, logger *logrus.Entry)
 	return tablesWithData, emptyTables, nil
 }
 
+// removeInvisibleIndeces removes all invisible idx references from a go_mysql.schema.Table.
+func removeInvisibleIndeces(ts *schema.Table) {
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply table_schema_cache.go]: Done.

File: dml_events.go [hunk 5]
```diff
@@ -347,53 +354,62 @@ func verifyValuesHasTheSameLengthAsColumns(table *TableSchema, values ...RowData
...
 
-func buildStringListForValues(columns []schema.TableColumn, values []interface{}) string {
+func buildStringListForValues(table *TableSchema, values []interface{}) string {
 	var buffer []byte
 
 	for i, value := range values {
-		if i > 0 {
+		if table.IsColumnIndexGenerated(i) {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply dml_events.go]: > why do we recheck already filtered out values?

Ah, this is a leftover from the first approach :facepalm: Fixed.

This should now work fine, regardless of column ordering.

File: row_batch.go [hunk 6]
```diff
@@ -64,23 +64,27 @@ func (e *RowBatch) AsSQLQuery(schemaName, tableName string) (string, []interface
...
-	valuesStr := "(" + strings.Repeat("?,", len(e.columns)-1) + "?)"
+	filteredColumns := e.table.NonGeneratedColumnNames()
+
+	valuesStr := "(" + strings.Repeat("?,", len(filteredColumns)-1) + "?)"
 	valuesStr = strings.Repeat(valuesStr+",", len(e.values)-1) + valuesStr
 
 	query := "INSERT IGNORE INTO " +
 		QuotedTableNameFromString(schemaName, tableName) +
-		" (" + strings.Join(QuoteFields(e.columns), ",") + ") VALUES " + valuesStr
+		" (" + strings.Join(filteredColumns, ",") + ") VALUES " + valuesStr
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply row_batch.go]: Yikes, good call, i somehow ate up the `QuoteFields()` call in here. Fixed.

File: dml_events.go [hunk 7]
```diff
@@ -281,16 +282,22 @@ func (e *BinlogDeleteEvent) PaginationKey() (string, error) {
...
 				len(table.Columns),
-				len(row),
+				len(rawRow),
 			)
 		}
+
+		row, err := table.FilterGeneratedColumnsOnRowData(rawRow)
+		if err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (proton-lisandro-pin) [reply dml_events.go]: > this doesn't make sense for me, we return copy of the row without generated columns, what for? wouldn't it be simpler to skip is generated in the following for loop?

Cleanliness, mainly.

  * `FilterGeneratedColumnsOnRowData()` now encapsulates all checks related to row data and column data matching, both pre- and post-filtering, so we don't repeat it all over.
  *  _Not_ doing filtering within `NewBinlogDMLEvents()` means that we'd have to repeat that logic for every single `New*Events()` called right after, which seems kinda brittle.

**Comment by driv3r** [comment]
@proton-lisandro-pin lets extend the tests to include those - they exercise some edge cases, and lets rebase the branch, lets wrap it up this week. 

```diff
diff --git a/test/go/dml_events_test.go b/test/go/dml_events_test.go
index d0b2cec..4bcb7f3 100644
--- a/test/go/dml_events_test.go
+++ b/test/go/dml_events_test.go
@@ -390,6 +390,163 @@ func (this *DMLEventsTestSuite) TestNoRowsQueryEvent() {
 	this.Require().Equal("", annotation)
 }
 
+// TestNewBinlogDMLEventsUnsignedConversionWithGeneratedColumn exercises two bugs
+// that arise when a virtual column sits before an unsigned integer column.
+//
+// Bug 1 (panic / index out of range): the second commit of the PR compacts the
+// raw row by removing generated column values BEFORE applying the unsigned-integer
+// normalisation loop.  The loop still iterates over table.Columns (full length),
+// so row[i] for the unsigned column indexes into the shortened slice and panics.
+//
+// Bug 2 (wrong value): if the panic is suppressed or the generated column happens
+// to be last, the unsigned normalisation is applied to the wrong row position,
+// leaving int8(-1) serialised as -1 instead of the correct uint8(255).
+func (this *DMLEventsTestSuite) TestNewBinlogDMLEventsUnsignedConversionWithGeneratedColumn() {
+	columns := []schema.TableColumn{
+		{Name: "id"},
+		{Name: "gen", IsVirtual: true},
+		{Name: "u8", IsUnsigned: true},
+	}
+	table := &ghostferry.TableSchema{
+		Table: &schema.Table{
+			Schema:  "test_schema",
+			Name:    "test_table",
+			Columns: columns,
+		},
+	}
+
+	ev := &replication.BinlogEvent{
+		Header: &replication.EventHeader{EventType: replication.WRITE_ROWS_EVENTv2},
+		Event: &replication.RowsEvent{
+			Rows: [][]interface{}{
+				{int64(1000), "gen_val", int8(-1)},
+			},
+		},
+	}
+
+	dmlEvents, err := ghostferry.NewBinlogDMLEvents(table, ev, mysql.Position{}, mysql.Position{}, nil)
+	this.Require().Nil(err)
+	this.Require().Equal(1, len(dmlEvents))
+
+	q, err := dmlEvents[0].AsSQLString("test_schema", "test_table")
+	this.Require().Nil(err)
+	this.Require().Equal(
+		"INSERT IGNORE INTO `test_schema`.`test_table` (`id`,`u8`) VALUES (1000,255)",
+		q,
+	)
+}
+
+// TestBinlogInsertEventGeneratedColumnBeforeJSONPreservesJSONCasting exercises
+// the metadata misalignment in buildStringListForValues introduced by the PR.
+//
+// AsSQLString filters generated column values out of the row before passing the
+// shortened slice to buildStringListForValues.  That function then uses the loop
+// counter i to index table.Columns, so the JSON column's value (at position 0
+// in the filtered slice) is looked up against table.Columns[0] — the virtual
+// column — which has no JSON type.  As a result the value is emitted as a plain
+// escaped string instead of CAST(... AS JSON).
+func (this *DMLEventsTestSuite) TestBinlogInsertEventGeneratedColumnBeforeJSONPreservesJSONCasting() {
+	columns := []schema.TableColumn{
+		{Name: "gen", IsVirtual: true},
+		{Name: "payload", Type: schema.TYPE_JSON},
+	}
+	table := &ghostferry.TableSchema{
+		Table: &schema.Table{
+			Schema:  "test_schema",
+			Name:    "test_table",
+			Columns: columns,
+		},
+	}
+	eventBase := ghostferry.NewDMLEventBase(table, mysql.Position{}, mysql.Position{}, nil, time.Unix(1618318965, 0))
+
+	rowsEvent := &replication.RowsEvent{
+		Table: this.tableMapEvent,
+		Rows:  [][]interface{}{{"gen_val", []byte("payload_data")}},
+	}
+
+	dmlEvents, err := ghostferry.NewBinlogInsertEvents(eventBase, rowsEvent)
+	this.Require().Nil(err)
+	this.Require().Equal(1, len(dmlEvents))
+
+	q, err := dmlEvents[0].AsSQLString("test_schema", "test_table")
+	this.Require().Nil(err)
+	this.Require().Equal(
+		"INSERT IGNORE INTO `test_schema`.`test_table` (`payload`) VALUES (CAST('payload_data' AS JSON))",
+		q,
+	)
+}
+
+// TestBinlogUpdateEventExcludesGeneratedColumnFromSetAndWhere verifies that
+// UPDATE events for tables with virtual columns emit SET and WHERE clauses that
+// reference only real (non-generated) columns.
+func (this *DMLEventsTestSuite) TestBinlogUpdateEventExcludesGeneratedColumnFromSetAndWhere() {
+	columns := []schema.TableColumn{
+		{Name: "id"},
+		{Name: "gen", IsVirtual: true},
+		{Name: "data"},
+	}
+	table := &ghostferry.TableSchema{
+		Table: &schema.Table{
+			Schema:  "test_schema",
+			Name:    "test_table",
+			Columns: columns,
+		},
+	}
+	eventBase := ghostferry.NewDMLEventBase(table, mysql.Position{}, mysql.Position{}, nil, time.Unix(1618318965, 0))
+
+	rowsEvent := &replication.RowsEvent{
+		Table: this.tableMapEvent,
+		Rows: [][]interface{}{
+			{int64(1000), "gen_old", "old_data"},
+			{int64(1000), "gen_new", "new_data"},
+		},
+	}
+
+	dmlEvents, err := ghostferry.NewBinlogUpdateEvents(eventBase, rowsEvent)
+	this.Require().Nil(err)
+	this.Require().Equal(1, len(dmlEvents))
+
+	q, err := dmlEvents[0].AsSQLString("test_schema", "test_table")
+	this.Require().Nil(err)
+	this.Require().Equal(
+		"UPDATE `test_schema`.`test_table` SET `id`=1000,`data`='new_data' WHERE `id`=1000 AND `data`='old_data'",
+		q,
+	)
+}
+
+// TestBinlogDeleteEventExcludesStoredGeneratedColumnFromWhere verifies that
+// DELETE events skip both VIRTUAL and STORED generated columns in the WHERE clause.
+func (this *DMLEventsTestSuite) TestBinlogDeleteEventExcludesStoredGeneratedColumnFromWhere() {
+	columns := []schema.TableColumn{
+		{Name: "id"},
+		{Name: "data"},
+		{Name: "summary", IsStored: true},
+	}
+	table := &ghostferry.TableSchema{
+		Table: &schema.Table{
+			Schema:  "test_schema",
+			Name:    "test_table",
+			Columns: columns,
+		},
+	}
+	eventBase := ghostferry.NewDMLEventBase(table, mysql.Position{}, mysql.Position{}, nil, time.Unix(1618318965, 0))
+
+	rowsEvent := &replication.RowsEvent{
+		Table: this.tableMapEvent,
+		Rows:  [][]interface{}{{int64(1000), "hello", "abc123"}},
+	}
+
+	dmlEvents, err := ghostferry.NewBinlogDeleteEvents(eventBase, rowsEvent)
+	this.Require().Nil(err)
+	this.Require().Equal(1, len(dmlEvents))
+
+	q, err := dmlEvents[0].AsSQLString("test_schema", "test_table")
+	this.Require().Nil(err)
+	this.Require().Equal(
+		"DELETE FROM `test_schema`.`test_table` WHERE `id`=1000 AND `data`='hello'",
+		q,
+	)
+}
 func TestDMLEventsTestSuite(t *testing.T) {
 	suite.Run(t, new(DMLEventsTestSuite))
 }
```

**Comment by driv3r** [comment]
lets wrap it up in #437 check it out for the CLA thing, as I've pulled in your commits 👍 unless you prefer to continue on your own, then it's also fine :)

**Comment by proton-lisandro-pin** [comment]
> @proton-lisandro-pin lets extend the tests to include those - they exercise some edge cases, and lets rebase the branch, lets wrap it up this week.

Apologies for the late reply, i've been OOO for the last couple days.

Just pushed a fix for the issues with column/value index mismatch, and added your suggested tests (thanks!). These seem to pass fine - and the overall process is benchmarking a bit faster than the previous iteration as well.
